File: #49_Anagrams/Solution.py
class Solution:
	# @param {string[]} strs
	# @return {string[]}
	def anagrams(self, strs):
		if not strs:
			return []
		re = []
		map = {}
		for s in strs:
			ss = self.sortStr(s) # sorted string
			map[ss] = map.get(ss, 0) + 1
		for s in strs:
			if map[self.sortStr(s)] > 1:
				re.append(s)
		return re

	# radix sort, runs in linear time
	# @param {string} string
	# @return {string} hash value
	def sortStr(self, string):
		if not string:
			return string
		return ''.join(sorted(string))

sol = Solution()
print sol.sortStr("asp")
print sol.sortStr("jul")
# print sol.anagrams(["ray","cod","abe","ned","arc","jar","owl","pop","paw","sky","yup","fed","jul","woo","ado","why","ben","mys","den","dem","fat","you","eon","sui","oct","asp","ago","lea","sow","hus","fee","yup","eve","red","flo","ids","tic","pup","hag","ito","zoo"])

File: #49_Anagrams/Solution.java
import java.util.List;
import java.util.LinkedList;
import java.util.Arrays;
import java.util.HashMap;

public class Solution {
    public static List<String> anagrams(String[] strs) {
        List<String> result = new LinkedList<String>();
        if (strs.length == 0) return result;
        int len = strs.length;
        HashMap<String, Integer> map = new HashMap<String, Integer>();
        String[] sortedStr = new String[len];
        for (int i = 0; i < len; i++) {
        	String sorted = sort(strs[i]);
        	sortedStr[i] = sorted;
        	map.put(sorted, map.containsKey(sorted) ? (map.get(sorted) + 1) : 0);
        }
        for (int i = 0; i < len; i++) {
        	if (map.get(sortedStr[i]) > 0) result.add(strs[i]);
        }
        return result;
    }

    private static String sort(String str) {
    	char[] a = str.toCharArray();
    	Arrays.sort(a);
    	return String.valueOf(a);
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	String[] test = {"god", "dog", "ear", "are", "you", "pi", "ip", "string"};
    	List<String> list = sol.anagrams(test);
    	for (String str : list) {
    		System.out.println(str);
    	}
    }
}